### ISSUE TITLE
[Feature request] log/power transform in `DataProcessor`

### ISSUE THREAD

**Issue by patel-zeel** [tom-andersson/deepsensor#39]
As discussed in https://github.com/tom-andersson/deepsensor/issues/27#issuecomment-1684325388, some datasets may have a non-Gaussian distribution and it may help to bring the distribution close to Gaussian via a log or power transform. For example, an air quality dataset distribution looks like the following:
![image](https://github.com/tom-andersson/deepsensor/assets/59758528/ab13bbfb-63e7-4b22-a6b4-7a031040590f)

After the log and [power](https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.PowerTransformer.html) transform, the distribution changes to:
![image](https://github.com/tom-andersson/deepsensor/assets/59758528/dda02ec6-4590-400f-b61c-2eda751c296c)

An important point is to make sure a transform can later be inverted with the "change of variable" trick after predicting the mean and variance in the transformed space. The mean and the mode of the inverted distribution are no longer necessary to be the same.

**Comment by tom-andersson** [comment]
Thanks for opening this @patel-zeel :-) Sharing some thoughts below -  let me know if you'd be interested to take this on with a PR.

Adding log/power norm transforms for deterministic variables in the `DataProcessor` is simple. However, as you suggest, the main complication comes from having to unnormalise/invert the Gaussian distributions of the CNP by applying the `DataProcessor` to the mean and stddev from `DeepSensorModel.predict` here: https://github.com/tom-andersson/deepsensor/blob/d1ce7ac2e6d538d84d046d76d1335b94a22e01c7/deepsensor/model/model.py#L447-L448

DeepSensor currently assumes a linear mapping which preserves the Gaussian distribution, and so computing the unnormalised stddev is enabled by `add_offset=False` in `DataProcessor.map_array` (i.e. just scaling by `b` in `y=ax+b`). It would be tricky to support non-Gaussian distributions induced by non-linear transformations like power norm (although not out of the scope of DeepSensor in general). Much simpler would be to match the mean and stddev of the inverted distribution (assuming those can be computed easily using the change of variable trick) and have `DeepSensorModel.predict` just return a Gaussian with that matched mean/stddev.

In terms of the `DataProcessor`, only two methods should need updating: [DataProcessor.get_norm_params](https://github.com/tom-andersson/deepsensor/blob/d1ce7ac2e6d538d84d046d76d1335b94a22e01c7/deepsensor/data/processor.py#L164) and [DataProcessor.map_array](https://github.com/tom-andersson/deepsensor/blob/d1ce7ac2e6d538d84d046d76d1335b94a22e01c7/deepsensor/data/processor.py#L298-L352).

**Computing normalisation parameters**
Parameters for normalisation mappings are computed by [DataProcessor.get_norm_params](https://github.com/tom-andersson/deepsensor/blob/d1ce7ac2e6d538d84d046d76d1335b94a22e01c7/deepsensor/data/processor.py#L164) (if not pre-computed), and are stored in the `DataProcessor.norm_params` like this:
```
'air': {'method': 'mean_std',
         'param1': 281.255126953125,
         'param2': 16.320409774780273},
```

Note, the above config assumes there are only two parameters for a normalisation method - I've opened an issue to have a single `params` entry with a list of parameters (https://github.com/tom-andersson/deepsensor/issues/40).

For power norm, we can just `sklearn.preprocessing.PowerTransformer.fit` and then store the `.lambdas_` parameters (and whatever else we'd need to instantiate the same `PowerTransform` object in a new session). 

**Normalising/unnormalising data***
We would need a new entry in `DataProcessor.map_array` like the below: https://github.com/tom-andersson/deepsensor/blob/d1ce7ac2e6d538d84d046d76d1335b94a22e01c7/deepsensor/data/processor.py#L330-L340

Except we would instantiate a `PowerTransformer` object from `self.norm_params` and use `PowerTransformer.transform` (if `unnorm=False`) or `PowerTransformer.inverse_transform` (if `unnorm=True`).

We would then need to think about how to handle non-deterministic variables here (as discussed above), which may warrant replacing the `add_offset` kwarg with something more general that works with non-linear transformations.

cc @wesselb

**Comment by patel-zeel** [comment]
Thanks for the headstart, @tom-andersson! I'd like to take on this PR and will share my thoughts next week once I am free from my 6 monthly Ph.D. evaluations.

**Comment by patel-zeel** [comment]
@tom-andersson, the following code with `sklearn.preprocessing.PowerTransformer` works for normalizing and unnormalizing with two separate objects:

```py
from sklearn.preprocessing import PowerTransformer, StandardScaler
import numpy as np

# dummy data (must be 2D)
x = np.array([[1.0, 2.0, 3.0]]).reshape(-1, 1)

# define scaler and fit
scaler = PowerTransformer()
x_scaled = scaler.fit_transform(x)
params = {
    "lambdas": scaler.lambdas_,
    "mean": scaler._scaler.mean_,
    "std": scaler._scaler.scale_,
}
print(params)

# redefine scaler and set scaler parameters from previous scaler
new_scaler = PowerTransformer()
new_scaler.lambdas_ = params["lambdas"]
new_scaler._scaler = StandardScaler()
new_scaler._scaler.mean_ = params["mean"]
new_scaler._scaler.scale_ = params["std"]

# inverse transform data
x_rescaled = new_scaler.inverse_transform(x_scaled)

# check if inverse transform is correct
assert np.allclose(x, x_rescaled)
```

the `_yeo_johnson_transform` in sklearn is here: https://github.com/scikit-learn/scikit-learn/blob/7f9bad99d6e0a3e8ddf92a7e5561245224dab102/sklearn/preprocessing/_data.py#L3307C25-L3316 here. 

Trying to work out some math:
The transformation for $\lambda > 0$ is $y = f(x) = \frac{(x + 1)^{\lambda}}{\lambda}$.
The inverse transformation: $x = f^{-1}(y) = \left(\lambda y + 1\right)^{\frac{1}{\lambda}} - 1$.

At prediction time, given $y \sim \mathcal{N}(\mu, \sigma^2)$, it seems hard to find analytical mean and variance of $p(x)$ in terms of $\mu$ and $\sigma$.

**Comment by tom-andersson** [comment]
Hey @patel-zeel, nice work on reinstantiating the `PowerTransformer` from the parameters! However, if we can't compute an analytical or approximate mean/stddev from the inverse transform that would be a deal breaker. It does seem tricky - I think we would need Wolfram Alpha or some solver to compute the mean/stddev of the (non-Gaussian) transformed distribution.

I'm also not fully convinced non-linear transformations like this are worth the effort - it's not obvious to me that more Gaussian inputs would yield a significant performance improvement. Furthermore, the conditional PM2.5 distributions (ie when you condition on other observations and auxiliary data) will not look as non-Gaussian as when you simply marginalise over everything (as in the figure in your OP), so a Gaussian likelihood should be reasonable for PM2.5.

**Comment by tom-andersson** [comment]
Hey @patel-zeel, what is your opinion on the above? Do you think we should close this issue for now?

**Comment by patel-zeel** [comment]
I have mixed opinions on this and would love your thoughts on them. I think just allowing `log1p` transform on data and inverse transform of `expm1` would be possible to implement since the resultant distribution becomes `LogNormal` and it has closed form mean and variance:

1. Currently, we don't have empirical evidence (at least I have not experimented with it specifically) suggesting that such transforms will help while using `deepsensor`.
2. In contrast, some users may find it a useful transform for their application, and it may also work as an initial example to see if `deepsensor` can incorporate other similar transforms. So, if it seems reasonable, we can simply implement the `log1p` transform and hope that someone may end up using it and suggest some modifications if needed.

In general, would it be helpful if `deepsensor` returns a distribution object (`torch.distributions` or `tensorflow_probabilisty.distributions`) instead of `mean` and `variance` to generelize beyond normality assumption? I personally find `xarray` versions of mean and variance very useful so there could be a utility function that takes an array of size test data (e.g. mean, variance, mode) and converts it into an `xarray` dataset. Something like:

```py
from deepsensor.post_processing import get_xarray

pred_dist = model.predict(test_task, X_t=ds_raw)
mean_ds = get_xarray(pred_dist.mean, data_processor)
mode_ds = get_xarray(pred_dist.mode, data_processor)
std_ds = get_xarray(pred_dist.stddev, data_processor)
```

**Comment by tom-andersson** [comment]
Closing for now, as discussed in our meeting last week. In summary, our thoughts are:
* Having explicitly user-chosen non-Gaussian output distributions from the `ConvNP` would definitely be useful.
* However, non-linear input transformations (which indirectly induce non-Gaussian output distributions) are too complicated and it's not clear we would get performance boosts by having the normalised input data be globally more Gaussian.
